test-adapter/src/main/scala/org/scalajs/sbttestadapter/TestAdapter.scala

```diff
@@ -0,0 +1,191 @@
+\*                                                                      */
+
+
+package org.scalajs.testadapter
+
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.collection.concurrent.TrieMap
+
+import java.util.concurrent.atomic.{AtomicInteger, AtomicBoolean}
+
+import org.scalajs.core.ir.Utils.escapeJS
+import org.scalajs.core.tools.io._
+import org.scalajs.core.tools.logging._
+import org.scalajs.core.tools.linker.ModuleKind
+
+import org.scalajs.jsenv._
+import org.scalajs.testcommon._
+
+import sbt.testing.Framework
+
+final class TestAdapter(jsEnv: ComJSEnv, config: TestAdapter.Config) {
+
+  import TestAdapter.ManagedRunner
+
+  /** Map of ThreadId -> ManagedRunner */
+  private[this] val runners = TrieMap.empty[Long, ManagedRunner]
+
+  private[this] val closing = new AtomicBoolean(false)
+
+  private[this] val nextRunID = new AtomicInteger(0)
+  private[this] val runs = TrieMap.empty[RunMux.RunID, Unit]
+
+  /** Creates an `sbt.testing.Framework` for each framework that can be found.
+   *
+   *  The returned Frameworks bind to this TestAdapter and are only valid until
+   *  [[close]] is called.
+   */
+  def loadFrameworks(frameworkNames: List[List[String]]): List[Option[Framework]] = {
+    val runner = getRunnerForThread()
+
+    val frameworks = runner.com
+      .call(JSEndpoints.detectFrameworks)(frameworkNames)
+      .map(_.map(_.map(info => new FrameworkAdapter(info, this))))
+
+    // If there is no testing framework loaded, nothing will reply.
+    val fallback: Future[List[Option[Framework]]] =
+      runner.runner.future.map(_ => frameworkNames.map(_ => None))
+
+    Future.firstCompletedOf(List(frameworks, fallback)).await()
+  }
+
+  /** Releases all resources. All associated runs must be done. */
+  def close(): Unit = {
+    if (!closing.getAndSet(true)) {
+      // Snapshot runs.
+      val seenRuns = runs.keySet
+
+      runners.values.foreach(_.com.close())
+      runners.values.foreach(_.runner.stop())
+      runners.clear()
+
+      runs.clear()
+      if (seenRuns.nonEmpty) {
+        throw new IllegalStateException(
+            s"close() called with incomplete runs: $seenRuns")
+      }
+    }
+  }
+
+  private[testadapter] def runStarting(): RunMux.RunID = {
+    require(!closing.get(), "We are closing. Cannot create new run.")
+    val runID = nextRunID.getAndIncrement()
+    runs.put(runID, ())
+    runID
+  }
+
+  /** Called by [[RunnerAdapter]] when the run is completed. */
+  private[testadapter] def runDone(runID: RunMux.RunID): Unit = {
+    val old = runs.remove(runID)
+    require(old.nonEmpty, s"Tried to remove nonexistent run $runID")
+  }
+
+  private[testadapter] def getRunnerForThread(): ManagedRunner = {
+    // Prevent runners from being started after closing started.
+    // Otherwise we might leak runners.
+    require(!closing.get(), "We are closing. Cannot create new runner.")
```

Comment:
What happens if `close()` is called right after this `require` has been tested? It seems that in that case we would be silently leaking stuff.

Reply:
Yes, indeed. This require is just to catch bugs, it doesn't actually enforce the synchronization requirement. Creating a run on a returned `Framework` or calling `close` when there are incomplete runs is undefined behavior.

I can of course make this threadsafe, but I'll probably need a lock, which I'd rather avoid here (since this method will be called for each test case, so I'd like to minimize potential contention).

Reply:
OK, that sounds reasonable.